File: frontend/lib/src/features/action/domain/action.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import '../../dimension/domain/dimension.dart';

part 'action.freezed.dart';
part 'action.g.dart';

@freezed
class Action with _$Action {
  const factory Action({
    required String id,
    required DateTime startTime,
    DateTime? endTime,
    String? description,
    @Default("Dovere") String category,
    @Default(3) int difficulty,
    required int fulfillmentScore,
    required String userId,
    required String dimensionId,
    Dimension? dimension,
  }) = _Action;

  factory Action.fromJson(Map<String, dynamic> json) => _$ActionFromJson(json);
}

@freezed
class ActionCreate with _$ActionCreate {
  const factory ActionCreate({
    String? description,
    DateTime? startTime,
    DateTime? endTime,
    required String dimensionId,
    required int fulfillmentScore,
    @Default("Dovere") String category,
    @Default(3) int difficulty,
    @Default("COMPLETED") String status,
  }) = _ActionCreate;

  factory ActionCreate.fromJson(Map<String, dynamic> json) => _$ActionCreateFromJson(json);
}

File: backend/app/schemas/action.py
from sqlmodel import SQLModel
import uuid
from typing import Optional
from datetime import datetime
from pydantic import field_validator

from .base import TunableBaseModel
from .dimension import DimensionRead
from .validators import validate_xss_basic

# --- Schemi Action ---

class ActionBase(TunableBaseModel):
    description: Optional[str] = None
    category: str = "Dovere"
    difficulty: int = 3
    status: str = "COMPLETED"
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    dimension_id: Optional[str] = None # Now String
    fulfillment_score: int = 3

    @field_validator("description")
    def validate_desc(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        return validate_xss_basic(v)
        
    @field_validator("fulfillment_score")
    def validate_score(cls, v: int) -> int:
        if not (1 <= v <= 5):
            raise ValueError("Fulfillment score must be between 1 and 5")
        return v

class ActionCreate(ActionBase):
    dimension_id: str # Required (Slug)
    fulfillment_score: int # Required

class ActionUpdate(ActionBase):
    pass

class ActionRead(ActionBase):
    id: uuid.UUID
    user_id: uuid.UUID
    created_at: Optional[datetime] = None
    dimension: Optional[DimensionRead] = None
